### PR TITLE
Fix problems with structured Masked Columns

### LINKED ISSUE #9649: Cannot initialize structured MaskedColumn from structured MaskedColumn
A very strange bug, found by testing my scintillometry package with the new astropy: I can initialize a structured `MaskedColumn`, but it doesn't round-trip!  I'll try to figure out what's up, but, @taldcroft, let me know if you have any ideas...
```
from astropy.table import MaskedColumn

In [8]: mc = MaskedColumn([(1., 2.), (3., 4.)], mask=[(False, False), (False, False)], dtype='f8,f8')

In [9]: mc
Out[9]: 
<MaskedColumn dtype='void128' length=2>
(1.0, 2.0)
(3.0, 4.0)

In [10]: MaskedColumn(mc)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-e901348b3ba2> in <module>()
----> 1 MaskedColumn(mc)

/data/mhvk/venv/astropy-dev/lib/python3.7/site-packages/astropy-4.1.dev27003-py3.7-linux-x86_64.egg/astropy/table/column.py in __new__(cls, data, name, mask, fill_value, dtype, shape, length, description, unit, format, meta, copy, copy_indices)
   1263             # Coerce the fill_value to the correct type since `data` may be a
   1264             # different dtype than self.
-> 1265             fill_value = self.dtype.type(data.fill_value)
   1266         self.fill_value = fill_value
   1267 

ValueError: Can't cast from structure to non-structure, except if the structure only has a single field.
```

### PR DESCRIPTION
**Pull request by mhvk**
fixes #9649 

Also fix overkeen viewing as `NdarrayMixin`, which could lose a column's name.

Spent a long time trying to understand what went wrong in table creation from list of dict, only to realize in part it was that in testing we now turn all warnings into exceptions, which changes the behaviour. Sigh.

EDIT:
fixes https://github.com/mhvk/scintillometry/issues/130

### CODE REVIEW

**Comment by mhvk** [comment]
@taldcroft - I think `4.0` is appropriate, since this fixes a bug that breaks at least one downstream package - which I think is what a release candidate is meant for.

File: CHANGES.rst [hunk 1]
```diff
@@ -546,6 +546,9 @@ astropy.table
 - Ensured that inserting ``np.ma.masked`` (or any other value with a mask) into
   a ``MaskedColumn`` causes a masked entry to be inserted. [#9623]
 
+- Ensured that one can initialize a ``MaskedColumn`` with another ``MaskedColumn``
```
[the remark this answers is not in the document]
Comment (mhvk) [reply CHANGES.rst]: Yes, makes sense; done.

**Comment by mhvk** [comment]
@bsipocz - if I misunderstand the milestone labeling, let me know! Main goal is just to ensure this does get into `4.0` proper.

**Comment by bsipocz** [comment]
Yes, bug and docs fixes will still get into 4.0 proper. Normally at this point I prefer milestoning every non critical fix as 4.0.1 though, and if they got merged before the release is out, I'll just deal with their remilestoning and backporting. But that case they are not showing up as a blocker in the milestone.

Anyway, this is now approved, so everything is perfectly reasonable as is.

Thanks @mhvk!

**Pull request merged**